File: problems/sieve/sieve.cpp
#include <cstdint>
#include <cstdlib>

int sieve(int limit) {
  uint8_t *is_prime = new uint8_t[limit + 1];
  for (int i = 0; i <= limit; i++)
    is_prime[i] = 1;

  is_prime[0] = 0;
  is_prime[1] = 0;

  for (int p = 2; p * p <= limit; p++) {
    if (is_prime[p]) {
      for (int i = p * p; i <= limit; i += p)
        is_prime[i] = 0;
    }
  }

  int count = 0;
  for (int i = 0; i <= limit; i++) {
    if (is_prime[i])
      count++;
  }

  delete[] is_prime;
  return count;
}

int main(int argc, char **argv) {
  int limit = 999999 + argc;
  int total_primes = 0;
  for (int i = 0; i < 100; i++) {
    total_primes += sieve(limit);
  }
  if (total_primes == (78498 * 100))
    return 0;
  return 1;
}

File: problems/sieve/sieve.rs
struct Arena {
    data: Vec<u8>,
}

impl Arena {
    fn new(capacity: usize) -> Self {
        Arena { data: vec![0; capacity] }
    }
}

fn sieve(arena: &mut Arena, limit: i32) -> i32 {
    let is_prime = &mut arena.data[..(limit + 1) as usize];
    is_prime.fill(1);
    is_prime[0] = 0;
    is_prime[1] = 0;

    let mut p: i32 = 2;
    while p * p <= limit {
        if is_prime[p as usize] != 0 {
            let mut i = p * p;
            while i <= limit {
                is_prime[i as usize] = 0;
                i += p;
            }
        }
        p += 1;
    }

    let mut count: i32 = 0;
    for i in 0..=limit {
        if is_prime[i as usize] != 0 {
            count += 1;
        }
    }
    count
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let limit = 1000000;
    let mut arena = Arena::new((limit + 1) as usize);
    let mut total_primes = 0;
    for _ in 0..200 {
        total_primes += sieve(&mut arena, limit);
    }
    if total_primes == 78498 * 200 {
        std::process::exit(0);
    }
    std::process::exit(1);
}
